driver-core/src/main/com/mongodb/client/model/expressions/Expression.java

```diff
@@ -114,4 +116,9 @@ public interface Expression {
     <T extends Expression> MapExpression<T> isMapOr(MapExpression<T> other);
 
     StringExpression asString();
+
+    <R extends Expression> R passTo(Function<? super Expression, ? extends R> f);
+
+    <R extends Expression> R switchOn(Function<Branches, ? extends BranchesTerminal<? super Expression, ? extends R>> on);
```

Comment:
[I don't know what to do with this, probably - nothing]

I noticed that the following code compiles:

```java
of(1).switchOn(on -> on
        .eq(of(0), v -> of("true")))
```

but stops compiling if I add the default branch:

```java
of(1).switchOn(on -> on
        .eq(of(0), v -> of("true"))
        .defaults(v -> of("false")))
```

This is very confusing even to me, despite me having reviewed the code. The technical reason behind the error is that since `of(0)` produced `IntegerExpression`, `eq` returns `BranchesTerminal<IntegerExpression, ...>`, which makes `default` to expect `IntegerExpression` as its input. On the other hand, `switchOn` accepts `Function<Branches, ? extends BranchesTerminal<? super Expression, R>>`, i.e., `default` is guaranteed to be provided with an `Expression`, but not necessary `IntegerExpression`, which is what the IntelliJ's compiler unsuccessfully trying to tell us by saying _"no instance(s) of type variable(s) exist so that `Expression` conforms to `IntegerExpression`"_.

Now one may say _"but why doesn't the compiler think that `eq` produces `BranchesTerminal<IntegerExpression, ...>` in the first example that compiles?"_ To this my answer is - I have no idea, the answer is likely in the guts of the JLS type inference rules, and I know nothing about them.

The code can be fixed by either of the two following ways:

```java
of(1).switchOn(on -> on
        .eq((Expression) of(0), v -> of("true"))
        .defaults(v -> of("false")));
```

```java
of(1).switchIntegerOn(on -> on
        .eq(of(0), v -> of("true"))
        .defaults(v -> of("false")));
```

Reply:
This was not noticeable in the original implementation because it allowed usages that were too broad (not typesafe). Now, in the following, switchOn expects the "input" type (T) to be an Expression, but the `eq` forces an inference of IntegerEx. 

```
of(1).switchOn(on -> on
        .eq(of(0), v -> of("true"))
        .defaults(v -> of("false")))
```

But we should already know the type of the input, since we now have a variety of methods to infer the type from. So I have added a T to Branches, which I should have done when I added those various switchOn methods.